### ISSUE TITLE
[sparse] Wrong jacobian computation for bcoo_spdot_general

### ISSUE THREAD

**Issue by lucagrementieri** [google/jax#10163]
It seems that the Jacobian computed by `bcoo_spdot_general`, invoked through the operator `@` does not organize the values in the correct order.

In order to show the bug, I have written the following code.
```python
import jax
import jax.numpy as jnp
from jax.experimental.sparse.bcoo import BCOO

A_indices = jnp.array([[0, 1], [0, 2], [1, 1], [1, 2], [1, 0]])
A_values = jnp.array([-2.0, 1.0, -1.0, 0.5, 2.0])
A_shape = (2, 3)

B_indices = jnp.array([[0, 2], [2, 1], [0, 3], [1, 3], [1, 0], [0, 0]])
B_values = jnp.array([10.0, 100.0, 1000.0, -5.0, -50.0, -500.0])
B_shape = (3, 4)

def sp_sp_product(v1, v2):
    sparse1 = BCOO((v1, A_indices), shape=A_shape)
    sparse2 = BCOO((v2, B_indices), shape=B_shape)
    return (sparse1 @ sparse2).todense()

def sp_dense_product(v1, v2):
    sparse1 = BCOO((v1, A_indices), shape=A_shape)
    dense2 = BCOO((v2, B_indices), shape=B_shape).todense()
    return sparse1 @ dense2

sp_sp_product_jac = jax.jacfwd(sp_sp_product, argnums=1)(A_values, B_values)
sparse_dense_product_jac = jax.jacfwd(sp_dense_product, argnums=1)(A_values, B_values)

assert jnp.array_equal(
    sp_sp_product_jac, sparse_dense_product_jac
), "The two Jacobian should be the same"
```

`sparse_dense_product_jac` is correct (as it can be verified analitically), while `sp_sp_product_jac` is incorrect since it should be identical to `sparse_dense_product_jac` but it is not.
The shape of `sp_sp_product_jac` in the expected one and the values in `sp_sp_product_jac` seems correct but they are not placed in the expected location. Probably there is a problem in the indices of the sparse jacobian matrix.

**Comment by lucagrementieri** [comment]
@jakevdp As the author of https://github.com/google/jax/pull/8848, have you any thought about it?

**Comment by jakevdp** [comment]
I'll take a look

**Comment by jakevdp** [comment]
Still working on debugging this... it's a tricky one.

**Comment by lucagrementieri** [comment]
Inspecting the code it seems that the problem is that the indices relative to the output JVP are assumed to be the indices of the output of the result of the product `bcoo_spdot_general`.

I have tried to return these indices modifying the function like that.

```python
def _bcoo_spdot_general_jvp(primals, tangents, **kwds):
    lhs_data, lhs_indices, rhs_data, rhs_indices = primals
    lhs_data_dot, lhs_indices_dot, rhs_data_dot, rhs_indices_dot = tangents
    primals_out = _bcoo_spdot_general(*primals, **kwds)
    assert type(lhs_indices_dot) is ad.Zero
    assert type(rhs_indices_dot) is ad.Zero
    if type(lhs_data_dot) is not ad.Zero:
        data_dot_out, indices_dot_out = _bcoo_spdot_general(
            lhs_data_dot, lhs_indices, rhs_data, rhs_indices, **kwds
        )
    if type(rhs_data_dot) is not ad.Zero:
        data_dot_out, indices_dot_out = _bcoo_spdot_general(
            lhs_data, lhs_indices, rhs_data_dot, rhs_indices, **kwds
        )
    return primals_out, [data_dot_out, indices_dot_out]
```

This code does not crash, but the tangent indices `indices_dot_out` are not passed to the JVP function of `bcoo_todense` (that makes the same assumption that tangent and primal data vectors share the same indices)
https://github.com/google/jax/blob/3184dd65a222354bffa2466d9a375162f5649132/jax/experimental/sparse/bcoo.py#L256
so the final result is identical.

**Comment by jakevdp** [comment]
The assumption that the tangent and primal vectors share the same indices is fundamental to how autodiff of sparse objecst works. I'm not sure that your modified function is the correct approach, because the gradient of integer indices is not those indices themselves, but a symbolic zero.

**Comment by jakevdp** [comment]
OK, I think the issue comes down to this line in the spdot_general implementation: https://github.com/google/jax/blob/5522ed1702de3e806ecb41b7adc00f9841e27a62/jax/experimental/sparse/bcoo.py#L1090

`sum_duplicates` sorts the indices and moves explicit zero-values to the end - since explicit zeros are different in each batch, it has to return a different set of indices for each batch, such that the output data no longer shares indices. We need a version of `sum_duplicates` that works correctly when data is batched and indices are not.

**Comment by jakevdp** [comment]
Put another way, this issue has brought up a more general concern that I hadn't considered before: sparse autodiff is fundamentally incompatible with any operation in which the output indices depend on the values in the data, because that may break the property that primals and tangents must share indices.

**Comment by jakevdp** [comment]
Here's a shorter example showing how grad of `sum_duplicates` is incorrect if there are zeros in the data. I believe this is another effect of the root cause of this bug:
```python
from functools import partial
import jax
import jax.numpy as jnp
from jax.experimental import sparse

indices = jnp.array([[1], [2]])
data = jnp.array([0.0, 1.0])

def f(data, sum_duplicates = True):
  mat = sparse.BCOO((data, indices), shape=(3,))
  if sum_duplicates:
    mat = mat.sum_duplicates()
  return mat.todense().sum()

print(jax.grad(partial(f, sum_duplicates=True))(data))
# [0. 1.]
print(jax.grad(partial(f, sum_duplicates=False))(data))
# [1. 1.]
```

**Comment by lucagrementieri** [comment]
It seems to me that you changed your mind with respect to the cause in the last comment, but I reply to your previous comment sharing my current knowledge.

> My suspicion at the moment is that this has something to do with the interaction between batching and autodiff that happens within a jacobian... we end up with a sparse array with batched data and unbatched indices, and I think our test coverage is not all that extensive for that case.

The error can be reproduced for the JVP only, without the batching involved in the computation of the Jacobian.

```python
import jax
import jax.numpy as jnp
from jax.experimental.sparse.bcoo import BCOO

A_indices = jnp.array([[0, 1], [0, 2], [1, 1], [1, 2], [1, 0]])
A_values = jnp.array([-2.0, 1.0, -1.0, 0.5, 2.0])
A_shape = (2, 3)

B_indices = jnp.array([[0, 2], [2, 1], [0, 3], [1, 3], [1, 0], [0, 0]])
B_values = jnp.array([10.0, 100.0, 1000.0, -5.0, -50.0, -500.0])
B_shape = (3, 4)


def sp_sp_product(b_v):
    sparse1 = BCOO((A_values, A_indices), shape=A_shape)
    sparse2 = BCOO((b_v, B_indices), shape=B_shape)
    return (sparse1 @ sparse2).todense()


def sp_dense_product(b_v):
    sparse1 = BCOO((A_values, A_indices), shape=A_shape)
    dense2 = BCOO((b_v, B_indices), shape=B_shape).todense()
    return sparse1 @ dense2


first_row_vector = jnp.eye(B_values.size)[0]

_, sp_sp_product_jvp = jax.jvp(sp_sp_product, (B_values,), (first_row_vector,))
_, sp_dense_product_jvp = jax.jvp(sp_dense_product, (B_values,), (first_row_vector,))

assert jnp.array_equal(
    sp_sp_product_jvp, sp_dense_product_jvp
), "The two JVP should be the same"
```

**Comment by lucagrementieri** [comment]
> Put another way, this issue has brought up a more general concern that I hadn't considered before: sparse autodiff is fundamentally incompatible with any operation in which the output indices depend on the values in the data, because that may break the property that primals and tangents must share indices.

I totally agree with this!
In my opinion a workaround is to remove the constraint that the tangent vector of indices is Zero and use it to store the indices. I know that it is not mathematically correct, but if this choice stays hidden in the code maybe it causes no harm.

I'm going to write a PR to show in practice how that could work without a major refactor of the API.

**Comment by lucagrementieri** [comment]
> I'm not sure that your modified function is the correct approach, because the gradient of integer indices is not those indices themselves, but must be a symbolic zero.

You are right, this constraint is hard to circumvent because it is deeply integrated into JAX.
If have written a version of the code that does not causes the bug for the examples I posted, but it breaks in other tests linked to sparse matrices because when the tangent of the indices isn't zero, then an assertion fails in
https://github.com/google/jax/blob/0c02f7935aa90d1d895f34882186e889e2b8923e/jax/interpreters/ad.py#L454

I'm willing to help fix the problem, but I do not have a sufficient knowledge of JAX to come up with a solution that is consistent with the established inner workings of the library.

**Comment by jakevdp** [comment]
I don't think this will require any deep fix, we just need to not use a problematic version of `sum_duplicates` in the implementation. I think the solution will be to split the current `sum_duplicates` utility into a couple utilities, or perhaps one with several options: `sum_duplicates`, `sort_indices`, and `eliminate_zeros`. The first two are fine for autodiff, the third should error on autodiff. Then in functions that we want to be differentiable, we don't call the third.

**Comment by jakevdp** [comment]
#10198 fixes this issue; I'm going to address the larger issue of `sum_duplicates` and autodiff in a separate PR.
